### ISSUE TITLE
Idea - Possibility of parallel operation?

### ISSUE THREAD

**Issue by apricotpoodle** [spryker/cakephp-statemachine#33]
# Idea - Possibility of parallel operation?

Hello,

## I'd a quick look at your 

- documentation 
    - https://github.com/spryker/cakephp-statemachine/wiki
    - https://github.com/spryker/cakephp-statemachine/tree/master/docs

- video
    - https://www.youtube.com/watch?v=-lYdXw9oEZE
    - https://www.youtube.com/watch?v=pITF35FIjHk

## Question

- Is there, or will there be, a way to model processing, not sequential but parallel?

I have not seen this type of behavior example.

Have I misread, or misunderstood?

**Comment by dereuromark** [comment]
Parallel should be possible using multiple queues and thus processing streams
It should still be sequential for a specific item, but for different unrelated items it should be fine to run in parallel this way.

**Comment by apricotpoodle** [comment]
Thank you for your answer.

So for one specific item there is no way, at present time, to run multiple queues to get a parrallel behavior.

Anyway I'll continue to dig in the documentation, to try to understand queues and sub queues.

Or I'll have to change the strucutre of my tables.

One more time, thank you for your answer.

**Comment by dereuromark** [comment]
I dont understand how that would work, since a state machine graph is a chain of transitions
paralyzing this only creates conflicts in this chain, no?

**Comment by apricotpoodle** [comment]
Hello Mark

I will try to explain what I mean by a parallel operation of a finite state machine.

I’ll base my answer on the explanation given by Tony Marston on his site:
   
   -    
1st link

   -    
At the end of the email, I indicate a scenario where parallel operation is possible on the same record.


Workflow Modelling with Petri Nets

In order to implement a workflow system it is first necessary to find a suitable means of designing and modeling a workflow process. For this I take advantage of the work done by Carl Adam Petri who was the first to formulate a general theory for discrete parallel systems which gave birth to what are now known as Petri Nets.

Petri Nets is a formal and graphical language which is appropriate for modelling systems with concurrency and resource sharing. It is a generalisation of automata theory such that the concept of concurrently occurring events can be expressed.
Petri Nets have become so popular and widespread that there is a Platform Independent PetriNet Editor (PIPE) available, and it even has its own Petri Net Markup Language (PNML).

Objects within a Petri Net

The Petri Net language contains the following basic objects:

| Name | Description |
| Places | These are inactive and are analogous to inboxes in an office-based system. They are shown as circles in a Petri Net diagram. Each Petri Net has one start place and one end place, but any number of intermediate places. |
| Transitions | These are active and represent tasks to be performed. They are shown as rectangles in a Petri Net diagram. |
| Arcs | Each of these joins a single Place to a single Transition. They are shown as connecting lines in a Petri Net diagram. An inward arc goes from a Place to a Transition and an outward arc goes from a Transition to a Place. |
| Tokens | These represent the current state of a workflow process. They are shown as black dots within Places in a Petri Net diagram. A place can hold zero or more tokens at any moment in time. |


These objects are subject to the following rules: 
   
   - Places do nothing but hold Tokens representing the state of the process. A Place can hold zero or more Tokens at any moment in time. 
   - An Arc connects a Place to a Transition.       
      - Place P is called an input place of transition T if there exists a directed arc from P to T 
      - Place P is called an output place of transition T if there exists a directed arc from T to P. 

   - When an enabled transition is fired it moves tokens from all its input places to all its output places.       
      - Transition T is said to be enabled if each input place P of T contains at least one FREE token. 
      - How an enabled transition is actually fired depends on the type of trigger. 
      - When transition T is fired it consumes one token from each input place P of T and produces one token in each output place P of T. 

   - Each workflow process has a single start place. It must have at least one inward arc going into a transition. It may have an outward arc coming from a transition in order to restart the process. 
   - Each workflow process has a single end place. It must have at least one outward arc coming from a transition (it may have more than one), but it cannot have any inward arcs going into transitions. 

Routing within a Petri Net

The route between the start place and the end place within a workflow process can take several forms. 

These are:

| Routing Types |
| Sequential routing 
 |
|  |
| Parrallel routing 
 |
| Conditional routing 
 |
| Iterative routing 
 |


Splitting and Joining within a Petri Net

In order implement these routings you may employ a selection of splits and joins. 

These are: 
   
   - AND split 
   
   -    
   
      
      - An example of parallel routing where several tasks are performed in parallel or in no particular order. It is modeled by a transition with one input place and two or more output places. When fired the transition will create tokens in all output places. 

   
   - AND join       
      -       
      

      - A transition with two or more input places and one output place. This will only be enabled once there is a token in all of the input places, which would be after each parallel thread of execution has finished. 

   - Explicit OR split
   
   -    
   
        
      - An example of conditional routing where the decision is made as early as possible. It is modeled by attaching conditions or guards to the arcs going out of a transition to different places. 
      - Guard : An expression attached to an arc, shown in brackets, that evaluates to either TRUE or FALSE. Tokens can only travel over arcs when their guard evaluates to TRUE. The expression will typically involve the case attributes. 
      - More than two arcs of this type can come out of the same transition. They must all have a condition except the last arc as this will be used as the default path if none of the other conditions evaluates to TRUE. 

   
   - Implicit OR split       
      -       
      

      - An example of conditional routing where the decision is made as late as possible. Implicit or-splits are modeled as two or more arcs going from the same place but to different transitions. That way, the transition that happens to fire first will get the token. Once the token is gone, the remaining transitions are cancelled 
      - One of the transitions may have a timer as its trigger so that it will be fired if none of the other transitions is activated before the time limit expires. Expired transitions can either be triggered automatically via a background process which is running on a timer (e.g. cron), or manually via an online screen. 

   - OR join (explicit and implicit)       
      -       
      

      - Is simply a place that serves as the output place of two different transitions. That way, the next transition after the or-join place will be enabled when either of the two conditional threads are done. 


Example of a parallel workflow

Validation of a Journal article.
   
   - Spellchecking 
   - Editorial line validation 
   - launch of the proof 

Hire Validation String
   
   - Different users, or better, different user roles, each one on their ownn side, must either reject or accept a record (i.e. a hiring form). (i.e.) as soon as one role rejected it, the record is "deleted". Another refinement would be a guard timer would reject or accept it by default after a limit of time, one, or each one, of the role.

I hope all my logorhea is crystal clear.



​   Le mercredi 5 octobre 2022 à 14:42:57 UTC+2, Mark Scherer ***@***.***> a écrit :  
 
 


I dont understand how that would work, since a state machine graph is a chain of transitions
paralyzing this only creates conflicts in this chain, no?

**Comment by apricotpoodle** [comment]
hum the main link has disappered for m my previous answer. 

SO I copy it again . 
Main credits belongs to this site

[Radicore from Tony Marston](http://www.tonymarston.net/php-mysql/workflow.html#modelling)

**Comment by apricotpoodle** [comment]
IMHO parallel routing in statemachines, is well adapted for humantreatment, checking or validation of an item.

Still thinking of a Hiring Validation system.
Let's imagine than on the field a manager filla form for hiring someone,
then some N+1 or more levels Will have to validate or reject this demand.

In a sequential routing system, these human steps will last an eternity
if these humans don't check their daily todo list.

- In a parallel routing :
    - as soon the budget manager will say "NO! Too much expensive!" 
    - the rest of the validation is unuseful.

In the sequential routing system : 
- may be the budget checking step is not the first one 
    -  have to wait than each one of others humans has done his job.

**Comment by dereuromark** [comment]
I am fairly certain what you describe is out of scope of this plugin.
Feel free to make PRs if you feel otherwise. But I dont think there is capacity from our side to change things that dramatically right now.

**Comment by apricotpoodle** [comment]
I agree with you. anyway thank for your time. 
I'll try to use  this plugin anyway. 
I'll try the following strategy: Create a 'swarm' of small finite state machines, depending only on the role of the user (the verifiers). these little machines will simply have to answer a yes or no question and will have a single task to perform.
